### DISCUSSION TITLE
Truffleruby cannot build

### DISCUSSION THREAD

**Discussion by KieranP** [rbenv/ruby-build#1911] [Build failures] [answered]
### Steps to reproduce the behavior

```
% asdf install ruby truffleruby+graalvm-21.3.0
```

### Expected vs. actual behavior

Should install like other ruby interpreters, but does not.

### Logs

```
% asdf install ruby truffleruby+graalvm-21.3.0
ruby-build: using openssl from homebrew
Downloading graalvm-ce-java11-darwin-amd64-21.3.0.tar.gz...
-> https://dqw8nmjcqpjn7.cloudfront.net/6c2bf7f6e5fab901e8a2284a0dbec6ce214bde65aa80cfeb90bfef8eabb5f862
Installing truffleruby+graalvm-21.3.0...

ERROR: cannot install TruffleRuby to /Users/kieran/.asdf/installs/ruby/truffleruby+graalvm-21.3.0, which does not look like a valid TruffleRuby prefix
TruffleRuby only supports being installed to a not existing directory or replacing an existing TruffleRuby installation
See https://github.com/oracle/truffleruby/issues/1389 for details

BUILD FAILED (macOS 12.1 using ruby-build 20211227)

Inspect or clean up the working tree at /var/folders/xy/ryc1ry312wq8378j14qj7q100000gn/T/ruby-build.20220116211228.81181.SfKGkE
Results logged to /var/folders/xy/ryc1ry312wq8378j14qj7q100000gn/T/ruby-build.20220116211228.81181.log

Last 10 log lines:
ETag: "f89d77b6cfd089cca441e4cc6fe3422a-49"
Accept-Ranges: bytes
Server: AmazonS3
X-Cache: Hit from cloudfront
Via: 1.1 7f2634ab02503490a050083667b6ffd4.cloudfront.net (CloudFront)
X-Amz-Cf-Pop: ORD53-C3
X-Amz-Cf-Id: KJQpdLmn1fKQrYJc1RdPZDC8QbV8PHzlm98sdvpvsQ4w_RdwbdAcmw==
Age: 71794

/var/folders/xy/ryc1ry312wq8378j14qj7q100000gn/T/ruby-build.20220116211228.81181.SfKGkE/truffleruby+graalvm-21.3.0 /var/folders/xy/ryc1ry312wq8378j14qj7q100000gn/T/ruby-build.20220116211228.81181.SfKGkE ~
```

Even if I remove `/Users/kieran/.asdf/installs/ruby/truffleruby+graalvm-21.3.0`, the issue remains

**Comment by KieranP** [comment]
Bypassed asdf and ran ruby-build direct and it worked fine, so not a ruby-build issue

**Comment by pinealservo** [comment, marked as answer]
This is actually a `ruby-build` issue.  It was introduced here: https://github.com/rbenv/ruby-build/pull/1784

It's not inherent to the core of `ruby-build` but is due to code meant to prevent `ruby-build` from deleting `/usr/local` or other sorts of FHS-compliant install locations, which TruffleRuby doesn't support: https://github.com/oracle/truffleruby/issues/1389

The problem with the existing method of detecting these installs is that it's too general and conflicts with the way `asdf` install works. [This line](https://github.com/asdf-vm/asdf/blob/de6e22f909946f7d17047f4aeab41e581546b674/lib/commands/command-install.bash#L220) is run to make the directory just before the plugin (which in this case is `asdf-ruby` which then calls `ruby-build`) gets to run. These install directories are a core part of how asdf (a set of bash scripts) tracks what's installed and what versions are available.

It seems to me that the check should explicitly be for FHS paths rather than just existing directories, or at least they should explicitly allow existing paths with `.asdf` as a component in the path.

**Reply by eregon** [reply]
Is the directory created but empty when ruby-build is called?
I think we should allow that (it's no problem, doesn't remove any file), and find some way to check that in Bash somehow.
PR welcome, otherwise I'll do it when I get some time.

**Reply by pinealservo** [reply]
Yes, that seems to be the case. I thought of that option too a few minutes after creating the issue, and it definitely seems to avoid the original problem without depending on details of FHS or asdf's directory naming. I found some techniques for checking directory emptiness in bash, so I'll try taking a crack at a PR in a bit.

**Reply by eregon** [reply]
Fixed in https://github.com/rbenv/ruby-build/pull/1956
